### PR TITLE
feat!: support on cluster build from git repo with Tekton

### PR DESCRIPTION
**Pull request by zroubalik**
Signed-off-by: Zbynek Roubalik <zroubali@redhat.com>

# Changes
- :gift: support on cluster build from git repo with Tekton
- `func deploy --build` originally accepted `true` or `false`,  now it accepts `disabled`, `local` or `git` values

/kind enhancement

https://user-images.githubusercontent.com/726523/148761877-1561ae92-c530-44f3-b15f-aa96dff24bd2.mp4

Fixes:  https://github.com/knative-sandbox/kn-plugin-func/issues/748

### CODE REVIEW

**Comment by matejvasek** [comment]
@zroubalik please fix the e2e tests (`Error: specified build type "false" is not valid, allowed build types are "disabled", "local" or "git"`) and generated files `./hack/update-deps.sh`.

File: cmd/deploy.go [hunk 1]
```diff
@@ -198,10 +225,13 @@ func runDeploy(cmd *cobra.Command, _ []string, clientFn deployClientFn) (err err
 		return err
 	}
 
-	if config.Build {
+	switch function.BuildType {
```
Comment (matejvasek) [inline cmd/deploy.go:228]: Shouldn't we also handle `BuildTypeDisabled` or `default` case?

File: pipelines/tekton/pipeplines_provider.go [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+			pp.progressListener.Increment(fmt.Sprintf("Function deployed at URL: %s", ksvc.Status.URL.String()))
+		} else {
+			pp.progressListener.Increment(fmt.Sprintf("Function updated at URL: %s", ksvc.Status.URL.String()))
+		}
+	}
+
+	return nil
+}
+
+func (pp *PipelinesProvider) watchPipelineRunProgess(pr *v1beta1.PipelineRun) error {
```
Comment (matejvasek) [inline pipelines/tekton/pipeplines_provider.go:168]: typo

File: k8s/persistent_volumes.go [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+import (
+	"context"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func GetPersistentVolumeClaim(ctx context.Context, name, namespaceOverride string) (*corev1.PersistentVolumeClaim, error) {
```
Comment (matejvasek) [inline k8s/persistent_volumes.go:11]: Is this function used anywhere?

File: cmd/deploy.go (hunk 1 above)
Comment (zroubalik) [reply cmd/deploy.go:228]: It shouldn't happen here, but you are right, I will handle that in `default` case. Good catch.

File: k8s/persistent_volumes.go (hunk 3 above)
Comment (zroubalik) [reply k8s/persistent_volumes.go:11]: At the moment it is not, but most likely will be used in follow up code. I can remove it, if you prefer so.

File: k8s/role_bidings.go [hunk 4]
```diff
@@ -0,0 +1,64 @@
+package k8s
+
+import (
+	"context"
+
+	rbacv1 "k8s.io/api/rbac/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func CreateRoleBindingForSA(ctx context.Context, name, namespaceOverride, serviceAccountName, clusterRoleName string) (err error) {
```
[the remark this answers is not in the document]
Comment (zroubalik) [reply k8s/role_bidings.go:10]: leftover, good catch

File: k8s/persistent_volumes.go (hunk 3 above)
Comment (matejvasek) [reply k8s/persistent_volumes.go:11]: It's OK if it will be used soon.

**Comment by zroubalik** [comment]
> @zroubalik please fix the e2e tests (`Error: specified build type "false" is not valid, allowed build types are "disabled", "local" or "git"`) and generated files `./hack/update-deps.sh`.

Thanks for the review, I have fixed the problems.

Though `./hack/update-deps.sh` or ``./hack/update-codegen.sh` haven't done anything 🤷‍♂️ Not sure how to fix the validation problem.

**Comment by matejvasek** [comment]
> Though `./hack/update-deps.sh` or ``./hack/update-codegen.sh` haven't done anything :man_shrugging: Not sure how to fix the validation problem.

I think that @lkingland had similar issue on macOS.
For me the update removed this file: `third_party/VENDOR-LICENSE/github.com/opencontainers/selinux/go-selinux/LICENSE`.

**Comment by matejvasek** [comment]
@zroubalik also would it be possible to add some e2e test for in cluster build?

File: k8s/persistent_volumes.go [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+	client, namespace, err := NewClientAndResolvedNamespace(namespaceOverride)
+	if err != nil {
+		return nil, err
+	}
+
+	return client.CoreV1().PersistentVolumeClaims(namespace).Get(ctx, name, metav1.GetOptions{})
+}
+
+func CreatePersistentVolumeClaim(ctx context.Context, name, namespaceOverride string, accesMode corev1.PersistentVolumeAccessMode, resourceRequest resource.Quantity) (err error) {
```
Comment (matejvasek) [inline k8s/persistent_volumes.go:20]: isn't accesMode typo?

Comment (zroubalik) [reply k8s/persistent_volumes.go:20]: it is, fixing

**Comment by zroubalik** [comment]
> @zroubalik also would it be possible to add some e2e test for in cluster build?

Yeah, definitely. But I don't want to bloat this PR. I will create a few follow up issues(tasks) that need to be done for this feature.

**Comment by zroubalik** [comment]
> > Though `./hack/update-deps.sh` or ``./hack/update-codegen.sh` haven't done anything 🤷‍♂️ Not sure how to fix the validation problem.
> 
> I think that @lkingland had similar issue on macOS. For me the update removed this file: `third_party/VENDOR-LICENSE/github.com/opencontainers/selinux/go-selinux/LICENSE`.

@matejvasek I tried to remove that file manually, but then that codegen check is failing in the action, b/c the project is out of sync.

**Comment by matejvasek** [comment]
@zroubalik let me clean and re-run it.

**Comment by matejvasek** [comment]
@zroubalik try cherry-picking this: https://github.com/matejvasek/faas/commit/312b6e3b0cd83221fa3441e7e905f2081aaef827.

File: pipelines/tekton/pipeplines_provider.go [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+
+func (pp *PipelinesProvider) Run(ctx context.Context, f fn.Function) error {
+	pp.progressListener.Increment("Creating Pipeline resources")
+
+	client, err := NewTektonClient()
+	if err != nil {
+		return err
+	}
+
+	err = k8s.CreatePersistentVolumeClaim(ctx, getPipelinePvcName(f), pp.namespace, corev1.ReadWriteOnce, *resource.NewQuantity(500, resource.DecimalSI))
```
Comment (matejvasek) [inline pipelines/tekton/pipeplines_provider.go:80]: Is 500 bytes enough (I don't what is being stored there)?

File: k8s/secrets.go [hunk 7]
```diff
@@ -23,25 +19,89 @@ func GetSecret(ctx context.Context, name, namespaceOverride string) (*corev1.Sec
...
+	}
+	secret.Data[corev1.DockerConfigJsonKey] = dockerConfigJSONContent
+
+	_, err = client.CoreV1().Secrets(namespace).Create(ctx, secret, metav1.CreateOptions{})
 	return
 }
+
+// --- Helper methods for DockerConfigJson type of Secret
+// Taken from (and converted to private):
+// https://github.com/kubernetes/kubectl/blob/master/pkg/cmd/create/create_secret_docker.go
```
Comment (lance) [inline k8s/secrets.go:66]: Nit: Maybe useful to have a permalink here just in case?

File: k8s/secrets.go [hunk 8]
```diff
@@ -23,25 +19,89 @@ func GetSecret(ctx context.Context, name, namespaceOverride string) (*corev1.Sec
...
+		Auths: map[string]dockerConfigEntry{server: dockerConfigAuth},
+	}
+
+	return json.Marshal(dockerConfigJSON)
+}
+
+// encodeDockerConfigFieldAuth returns base64 encoding of the username and password string
+func encodeDockerConfigFieldAuth(username, password string) string {
+	fieldValue := username + ":" + password
+	return base64.StdEncoding.EncodeToString([]byte(fieldValue))
```
Comment (lance) [inline k8s/secrets.go:106]: I'm probably just missing it somewhere, but I didn't see where this is ever used/decoded.

File: pipelines/tekton/pipeplines_provider.go [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+		err := opt(pp)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return pp, nil
+}
+
+func (pp *PipelinesProvider) Run(ctx context.Context, f fn.Function) error {
```
Comment (lance) [inline pipelines/tekton/pipeplines_provider.go:72]: Some descriptive comments here might be helpful - in general, there is a lot going on in this file. I think comments throughout could be beneficial.

File: pipelines/tekton/pipeplines_provider.go [hunk 10]
```diff
@@ -0,0 +1,199 @@
...
+		return fmt.Errorf("problem in watching started pipeline run: %v", err)
+	}
+
+	pr, err = client.PipelineRuns(pp.namespace).Get(ctx, pr.Name, metav1.GetOptions{})
+	if err != nil {
+		return fmt.Errorf("problem in retriving pipeline run status: %v", err)
+	}
+
+	if pr.Status.GetCondition(apis.ConditionSucceeded).Status == corev1.ConditionFalse {
+		return fmt.Errorf("function pipeline run has failed, please inspect logs of Tekton PipelineRun \"%s\"", pr.Name)
```
Comment (lance) [inline pipelines/tekton/pipeplines_provider.go:145]: It would be really nice if we could display the last 20 or so lines of the log.

**Review by lance — COMMENTED** [review]
Really looking good! I have just a few small comments, @zroubalik

File: pipelines/tekton/pipeplines_provider.go (hunk 6 above)
Comment (lance) [reply pipelines/tekton/pipeplines_provider.go:80]: Isn't this where the function code is cloned?

File: k8s/role_bidings.go [hunk 11]
```diff
@@ -0,0 +1,36 @@
+package k8s
+
+import (
+	"context"
+
+	rbacv1 "k8s.io/api/rbac/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func CreateClusterRoleBindingForSA(ctx context.Context, name, namespaceOverride, serviceAccountName, clusterRoleName string) (err error) {
```
[the remark this answers is not in the document]
Comment (zroubalik) [reply k8s/role_bidings.go:10]: I didn't want to make it too long, but I think I can expand it the way you proposed.

File: function_git_unit_test.go [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+				URL: ptr.String("https://myrepo/foo.git"),
+			},
+			true,
+			0,
+		},
+		{
+			"correct 'Git - URL + revision",
+			Git{
+				URL:      ptr.String("https://myrepo/foo.git"),
+				Revision: ptr.String("mybranch"),
```
[the remark this answers is not in the document]
Comment (zroubalik) [reply function_git_unit_test.go:29]: You are right, I started with a test. We should add validation for those values as well.

File: k8s/secrets.go (hunk 8 above)
Comment (zroubalik) [reply k8s/secrets.go:106]: It's just a few lines above in the `handleDockerCfgJSONContent()` func.

**Comment by zroubalik** [comment]
> @zroubalik try cherry-picking this: [matejvasek@312b6e3](https://github.com/matejvasek/faas/commit/312b6e3b0cd83221fa3441e7e905f2081aaef827).

@matejvasek  seems like it helped, thanks a lot :)

File: pipelines/tekton/pipeplines_provider.go (hunk 6 above)
Comment (zroubalik) [reply pipelines/tekton/pipeplines_provider.go:83]: Yeah, it holds the source code and cache. It means `500 MiB` not `500 bytes` though :)

Comment (matejvasek) [reply pipelines/tekton/pipeplines_provider.go:83]: That weird.
```go
func TestPipelinesProvider_Run(t *testing.T) {
	var q *resource.Quantity
	q = resource.NewQuantity(500*(1_000_000), resource.DecimalSI)
	t.Logf("%+v", q)
	q = resource.NewQuantity(500, resource.DecimalSI)
	t.Logf("%+v", q)
}
```
Output:
```
=== RUN   TestPipelinesProvider_Run
    pipeplines_provider_test.go:11: 500M
    pipeplines_provider_test.go:13: 500
--- PASS: TestPipelinesProvider_Run (0.00s)
PASS
```
Maybe `k8s` multiplies the quantity by million? @zroubalik

Comment (matejvasek) [reply pipelines/tekton/pipeplines_provider.go:83]:
```go
	// Volume size, in bytes (e,g. 5Gi = 5GiB = 5 * 1024 * 1024 * 1024)
	ResourceStorage ResourceName = "storage"
```
really weird.

Comment (matejvasek) [reply pipelines/tekton/pipeplines_provider.go:83]:
```go
func TestPipelinesProvider_Run(t *testing.T) {

	pvc := &corev1.PersistentVolumeClaim{
		ObjectMeta: metav1.ObjectMeta{
			Name:      "nm",
			Namespace: "nms",
		},
		Spec: corev1.PersistentVolumeClaimSpec{
			AccessModes: []corev1.PersistentVolumeAccessMode{corev1.ReadWriteOnce},
			Resources: corev1.ResourceRequirements{
				Requests: corev1.ResourceList{},
			},
		},
	}
	pvc.Spec.Resources.Requests[corev1.ResourceStorage] = *resource.NewQuantity(500, resource.DecimalSI)
	enc := json.NewEncoder(os.Stderr)
	enc.Encode(&pvc)

	pvc.Spec.Resources.Requests[corev1.ResourceStorage] = *resource.NewQuantity(500*(1<<20), resource.BinarySI)
	enc.Encode(&pvc)
}
```
Ouptut:
```
{"metadata":{"name":"nm","namespace":"nms","creationTimestamp":null},"spec":{"accessModes":["ReadWriteOnce"],"resources":{"requests":{"storage":"500"}}},"status":{}}
{"metadata":{"name":"nm","namespace":"nms","creationTimestamp":null},"spec":{"accessModes":["ReadWriteOnce"],"resources":{"requests":{"storage":"500Mi"}}},"status":{}}
```

Comment (zroubalik) [reply pipelines/tekton/pipeplines_provider.go:83]: @matejvasek mabye it fallbacks to `Mi` if unit is not specified?

File: pipelines/tekton/pipeplines_provider.go [hunk 13]
```diff
@@ -0,0 +1,210 @@
...
+	} else {
```
Comment (lkingland) [inline pipelines/tekton/pipeplines_provider.go:149]: minor nit: the contents of this else statement can be moved out (the if statement returns)

File: pipelines/tekton/pipeplines_provider.go (hunk 6 above)
Comment (matejvasek) [reply pipelines/tekton/pipeplines_provider.go:83]: @rhuss @dsimansk any idea how this works?

Comment (zroubalik) [reply pipelines/tekton/pipeplines_provider.go:83]: I will explicitly specify the unit here, but the real question is whether 500Mi is ok value, it should store source code and buildpacks cache and layers.

**Review by lkingland — APPROVED** [review]
Looks really good, and I know this represents just the tip of the iceberg!

File: pipelines/tekton/pipeplines_provider.go (hunk 6 above)
Comment (zroubalik) [reply pipelines/tekton/pipeplines_provider.go]: I have added here the default size (500Mi) for now:
https://github.com/knative-sandbox/kn-plugin-func/blob/22923c455ed46da35ede04ce19346edd652788c7/pipelines/tekton/defaults.go#L4-L6
Should be cofigurable in the future.

File: pipelines/tekton/tasks.go [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+			Workspace: "source-workspace",
+		}},
+		Params: []pplnv1beta1.Param{
+			{Name: "url", Value: *pplnv1beta1.NewArrayOrString("$(params.gitRepository)")},
+			{Name: "revision", Value: *pplnv1beta1.NewArrayOrString("$(params.gitRevision)")},
+		},
+	}
+}
+
+func taskkBuild(runAfter string) pplnv1beta1.PipelineTask {
```
Comment (matejvasek) [inline pipelines/tekton/tasks.go:25]: typo?

**Comment by matejvasek** [comment]
When running build user is asked about registry URLbut isn't that known from `func.yaml`? @zroubalik

File: pipelines/tekton/tasks.go [hunk 15]
```diff
@@ -0,0 +1,116 @@
...
+}
+
+// TODO this should be part of the future func-build Tekton Task as a post-build step
+func taskImageDigest(runAfter string) pplnv1beta1.PipelineTask {
+	script := `#!/usr/bin/env bash
+               set -e
+
+			   func_file="/workspace/source/func.yaml"
+			   if [ "$(params.contextDir)" != "" ]; then
+			   		func_file = /workspace/source/$(params.contextDir)/func.yaml
```
Comment (matejvasek) [inline pipelines/tekton/tasks.go:56]: I think the spaces around the equal shouldn't be there.
Also it would be good tho have quotes around the value.

Comment (matejvasek) [reply pipelines/tekton/tasks.go:56]:
```diff
diff --git a/pipelines/tekton/tasks.go b/pipelines/tekton/tasks.go
index 16783524..20efc1d8 100644
--- a/pipelines/tekton/tasks.go
+++ b/pipelines/tekton/tasks.go
@@ -49,18 +49,19 @@ func taskkBuild(runAfter string) pplnv1beta1.PipelineTask {
 // TODO this should be part of the future func-build Tekton Task as a post-build step
 func taskImageDigest(runAfter string) pplnv1beta1.PipelineTask {
        script := `#!/usr/bin/env bash
-               set -e
 
-                          func_file="/workspace/source/func.yaml"
-                          if [ "$(params.contextDir)" != "" ]; then
-                                       func_file = /workspace/source/$(params.contextDir)/func.yaml
-                          fi
+set -e
 
-                          sed -i "s|^image:.*$|image: $(params.image)|" $func_file
-                          echo "Function image name: $(params.image)"
+func_file="/workspace/source/func.yaml"
+if [ "$(params.contextDir)" != "" ]; then
+  func_file="/workspace/source/$(params.contextDir)/func.yaml"
+fi
 
-                          sed -i "s/^imageDigest:.*$/imageDigest: $(params.digest)/" $func_file
-                          echo "Function image digest: $(params.digest)"
+sed -i "s|^image:.*$|image: $(params.image)|" "$func_file"
+echo "Function image name: $(params.image)"
+
+sed -i "s/^imageDigest:.*$/imageDigest: $(params.digest)/" "$func_file"
+echo "Function image digest: $(params.digest)"
        `
 
        return pplnv1beta1.PipelineTask{
```

**Comment by zroubalik** [comment]
> When running build user is asked about registry URL, but isn't that known from `func.yaml`? @zroubalik
> 
> Wouldn't it be possible to use existing credential provider? It could read credentials automatically form developer machine, but I am not sure whether it would be good thing.

@matejvasek I have captured both of these features in the Task list in https://github.com/knative-sandbox/kn-plugin-func/issues/620. 
Wrt registry url, yeah we should try to guess it from registry. And getting credentials automatically would be great, I am not sure if that's possible.

File: pipelines/tekton/tasks.go (hunk 15 above)
Comment (matejvasek) [reply pipelines/tekton/tasks.go:56]: possible improvement for future: using `yq` instead of `sed`.

**Comment by matejvasek** [comment]
>And getting credentials automatically would be great, I am not sure if that's possible.

Couldn't we use https://github.com/knative-sandbox/kn-plugin-func/blob/main/docker/creds/credentials.go#L153  @zroubalik ?

**Comment by zroubalik** [comment]
> > And getting credentials automatically would be great, I am not sure if that's possible.
> 
> Couldn't we use https://github.com/knative-sandbox/kn-plugin-func/blob/main/docker/creds/credentials.go#L153 @zroubalik ?

Probably yes :) Are you willing to implement this? If so, feel free to covert that task into an issue. It will be more than welcome :)

**Pull request merged**